File: SecureBlackbox Samples/PDF Encryptor/cpp/pdfencryptor.cpp
/*
 * SecureBlackbox 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../../include/secureblackbox.h"
#define LINE_LEN 100

namespace ArgParser {
    static char* optval(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc - 1; x++) {
            if (!strcmp(argv[x], option)) {
                return argv[x + 1];
            }
        }
        return (char*)"";
    }

    static bool optext(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc; x++) {
            if (!strcmp(argv[x], option)) {
                return true;
            }
        }
        return false;
    }
};

using namespace ArgParser;
PDFEncryptor pdfencryptor;
CertificateManager certificatemanager;

int showHelp() {
    printf("Usage:\n"
        "\t./pdfencryptor\t-inputFile <inputFile> -outputFile <outputFile>\n"
        "\t\t\t(-keyPath <path> -keyPassword <password> | -userPassword <password>) [-options ]?\n");
    printf("Options: \n");
    printf("\t-inputFile           (string) The PDF file to be encrypted.\n");
    printf("\t-outputFile          (string) Where to save the encrypted document.\n");
    printf("\t-keyPath             (string) The key file to be imported.\n");
    printf("\t-keyPassword         (string) The key password.\n");
    printf("\t-userPassword        (string) The password used for encryption.\n");
    printf("\t-encryptionAlgorithm (string) The encryption algorithm used to encrypt the document.\n");
    printf("\t-noMetadata          (switch) Specifies metadata should not be encrypted.\n");
    exit(1);
}

int main(int argc, char** argv) {
    if (argc < 7) {
        fprintf(stderr, "Error: Missing arguments.\n");
        showHelp();
    }

    optext(argc, argv, "-inputFile") ? pdfencryptor.SetInputFile(optval(argc, argv, "-inputFile")) : showHelp();
    optext(argc, argv, "-outputFile") ? pdfencryptor.SetOutputFile(optval(argc, argv, "-outputFile")) : showHelp();
    optext(argc, argv, "-encryptionAlgorithm")
        ? pdfencryptor.SetEncryptionAlgorithm(optval(argc, argv, "-encryptionAlgorithm"))
        : pdfencryptor.SetEncryptionAlgorithm("AES256");
    optext(argc, argv, "-noMetadata") ? pdfencryptor.SetEncryptMetadata(false) : pdfencryptor.SetEncryptMetadata(true);

    if (optext(argc, argv, "-keyPath") && optext(argc, argv, "-keyPassword")) {
        char* keyPath = optval(argc, argv, "-keyPath");
        char* keyPassword = optval(argc, argv, "-keyPassword");
        certificatemanager.ImportFromFile(keyPath, keyPassword);
        pdfencryptor.SetEncryptionType(PET_CERTIFICATE);
        pdfencryptor.SetEncryptionCertCount(1);
        pdfencryptor.SetEncryptionCertificateHandle(certificatemanager.GetCertHandle());
    } else if (optext(argc, argv, "-userPassword")) {
        char* password = optval(argc, argv, "-userPassword");
        pdfencryptor.SetEncryptionType(PET_PASSWORD);
        pdfencryptor.SetUserPassword(password);
    } else {
        showHelp();
    }

    const int retcode = pdfencryptor.Encrypt();

    if (retcode) {
        printf("Error [%d]: %s", pdfencryptor.GetLastErrorCode(), pdfencryptor.GetLastError());
        return 0;
    } else {
        printf("The document was encrypted successfully.\n");
    };

    fprintf(stderr, "\npress <return> to continue...\n");
    getchar();
    return 0;
}

File: SecureBlackbox Samples/PDF Encryptor/php/pdf_encryptor.php
<?php
/*
 * SecureBlackbox 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/secureblackbox_pdfencryptor.php');
require_once('../include/secureblackbox_certificatemanager.php');
require_once('../include/secureblackbox_const.php');
?>
<style>
table { width: 100% !important; }
td { white-space: nowrap; }
td input { width: 100%; }
td:last-child { width: 100%; }
</style>

<div width="90%">
  <form method=POST>
    <h2>PDF Encryptor Demo</h2>
    
    <h3>General Options</h3>
    <table>
      <tr><td>Input File:</td><td><input type=text name=inputFile value=""></td></tr>
      <tr><td>Output File:</td><td><input type=text name=outputFile value=""></td></tr>
    </table>
    <br/>
    <br/>

    <b>Encryption Options</b><br/><br/>
    <table>
      <tr>
        <td>Encryption algorithm:</td>
        <td>
          <select name="encAlg">
            <option value="RC440">RC4/40 bits (Acrobat 4)</option>
            <option value="RC4128">RC4/128 bits (Acrobat 5)</option>
            <option value="AES128">AES/128 bits (Acrobat 6, 7)</option>
            <option value="AES2569">AES/256 bits (Acrobat 9)</option>
            <option value="AES256X">AES/256 bits (Acrobat X)</option>
          </select>
        </td>
      </tr>
    </table>
    <br/>
    <label> <input type="checkbox" name="encryptMetadata" value="1" size="25"/> Encrypt document metadata </label>
    <br/>
    <br/>
    <br/>

    <b><label><input type ="radio" checked="checked" name="enctype" value="Password" /> Password encryption</label><b>
    <br/>
    <br/>
    <table>
      <tr><td>Password:</td><td><input type=password name=sPass value=""></td></tr>
    </table>
    <br/>
    <br/>

    <b><label><input type ="radio" name="enctype" value="Key" /> Public key encryption</label><b>
    <br/>
    <br/>
    <table>
      <tr><td>Certificate File:</td><td><input type=text name=sCertFile value=""></td></tr>
      <tr><td>Password:</td><td><input type=password name=sCertPass value=""></td></tr>
    </table>
    <br/>
    <br/>

    <input type="submit" value="Encrypt" />
  </form>
</div><br/>

<?php
  if ($_SERVER['REQUEST_METHOD'] == "POST") {
    $pdfencryptor = new SecureBlackbox_PDFEncryptor();
    $certmgr = new SecureBlackBox_CertificateManager();
    
    try {
      // General options
      $pdfencryptor->setInputFile($_REQUEST['inputFile']);
      $pdfencryptor->setOutputFile($_REQUEST['outputFile']);
      
      // Encryption options
      switch($_REQUEST['encAlg']){
        case "RC4128":
          $pdfencryptor->setEncryptionAlgorithm("RC4");
          $pdfencryptor->doConfig("RC4KeyBits=128");
          break;
        case "AES128":
          $pdfencryptor->setEncryptionAlgorithm("AES128");
          break;
        case "AES2569":
          $pdfencryptor->setEncryptionAlgorithm("AES256");
          $pdfencryptor->doConfig("HardenedKeyGeneration=false");
          break;
        case "AES256X":
          $pdfencryptor->setEncryptionAlgorithm("AES256");
          $pdfencryptor->doConfig("HardenedKeyGeneration=true");
          break;
        default:
          $pdfencryptor->setEncryptionAlgorithm("RC4");
          $pdfencryptor->doConfig("RC4KeyBits=40");
          break;
      }
      
      $encryptMetadata = (isset($_REQUEST['encryptMetadata']) && (($_REQUEST['encryptMetadata'] == 'yes') || ($_REQUEST['encryptMetadata'] == '1')));
      $pdfencryptor->setEncryptMetadata($encryptMetadata);
      
      $enctype = $_POST["enctype"];
      if ($enctype == "Password")
      {
        $pdfencryptor->setUserPassword($_REQUEST['sPass']);
        $pdfencryptor->setOwnerPassword($_REQUEST['sPass']);
      }
      else
      {
        $certmgr->doImportFromFile($_REQUEST['sCertFile'], $_REQUEST['sCertPass']);
        $pdfencryptor->setEncryptionCertificateHandle($certmgr->getCertHandle());
      }

      $pdfencryptor->doEncrypt();
      echo "<h2>PDF file successfully encrypted</h2>";
    }
    catch (exception $e) {
      echo "<h2>Encryption Failure (Details Below)</h2><p>" . $e->getMessage() . "</p>";
    }
  }
?>